Userland/Tests/Kernel/munmap-multi-region-unmapping.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Types.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+int main()
+{
+    {
+        printf("Testing full unnmap\n");
+        auto* map1 = mmap(nullptr, 2 * PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE | MAP_FIXED, 0, 0);
+        if (map1 == MAP_FAILED) {
+            perror("mmap 1");
+            return 1;
+        }
+        auto* map2 = mmap((void*)((FlatPtr)map1 + 2 * PAGE_SIZE), 2 * PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE | MAP_FIXED, 0, 0);
+        if (map2 == MAP_FAILED) {
+            perror("mmap 2");
+            return 1;
+        }
+
+        ((u32*)map1)[0] = 0x41414141;
+        ((u32*)map1)[PAGE_SIZE / sizeof(u32)] = 0x42424242;
+
+        ((u32*)map2)[0] = 0xbeefbeef;
+        ((u32*)map2)[PAGE_SIZE / sizeof(u32)] = 0xc0dec0de;
+
+        if (((u32*)map1)[0] != 0x41414141 || ((u32*)map1)[PAGE_SIZE / sizeof(u32)] != 0x42424242
+            || ((u32*)map2)[0] != 0xbeefbeef || ((u32*)map2)[PAGE_SIZE / sizeof(u32)] != 0xc0dec0de) {
+            perror("write");
+            return 1;
+        }
+
+        int res = munmap(map1, 4 * PAGE_SIZE);
+        if (res < 0) {
+            perror("unmap");
+            return 1;
+        }
+    }
+    {
+        printf("Testing partial unmaping\n");
+        auto* map1 = mmap(nullptr, 2 * PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE | MAP_FIXED, 0, 0);
```

Comment:
Does nullptr and `MAP_FIXED` actually work? That sounds like a bug in the Kernel, so let's not rely on it. (This affects the above `map1 = mmap(…` line, too.)

Reply:
I don't actually know how these parameters are supposed to work, but it runs without complaining

Reply:
The [man 2](https://man7.org/linux/man-pages/man2/mmap.2.html) page says that you should omit the MAP_FIXED flag,
but it also states, that MAP_FIXED can be really forcefull and overwrite old maps..

Reply:
Turns out, nullptr and MAP_FIXED triggers the `allocate_anywhere` branch, so this code is perfectly fine. I was wrong.

(Also, for future reference, our Kernel seems to reject trying to overwrite old maps.)